### ISSUE TITLE
Always going to about:blank at the end with 1.1.1 and Phantomjs 2.1.1

### ISSUE THREAD

**Issue by rhires** [casperjs/casperjs#1633]
This is on Windows 10 - I get the "about:blank" error. 
CasperJS version 1.1.1 at C:/node_modules/casperjs, using phantomjs version 2.1.1 

Example code:

```
var casper = require('casper').create({
    verbose: true,
    logLevel: 'debug',
    pageSettings: {
        loadImages: false,
        loadPlugins: false,
        userAgent: 'Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/39.0.2171.71 Safari/537.36 Edge/12.0'
    }
});

casper.on('remote.message', function(msg) {
    this.echo('remote message caught: ' + msg);
});

casper.on("page.error", function(msg,trace) {
    this.echo("Page Error: " + msg, "ERROR");
});

casper.start('https://www.google.fr');

casper.then(function(){
    console.log('Page Loaded');
    this.test.assertTitle('Google', 'Welcome to Google');
});

casper.run();
```

Example log output:

```
[info] [phantom] Starting...                                                                                                                                        
[info] [phantom] Running suite: 3 steps                                                                                                                             
[debug] [phantom] opening url: https://www.google.fr/, HTTP GET                                                                                                     
[debug] [phantom] Navigation requested: url=https://www.google.fr/, type=Other, willNavigate=true, isMainFrame=true                                                 
[debug] [phantom] url changed to "https://www.google.fr/"                                                                                                           
[debug] [phantom] Successfully injected Casper client-side utilities                                                                                                
[debug] [phantom] start page is loaded                                                                                                                              
[info] [phantom] Step anonymous 3/3 https://www.google.fr/ (HTTP 200)                                                                                               
Page Loaded                                                                                                                                                         
[debug] [phantom] Navigation requested: url=about:blank, type=Other, willNavigate=true, isMainFrame=true                                                            
[debug] [phantom] url changed to "about:blank"   
```

I've tried passing in some command line arguments from suggestions I've seen elsewhere, such as StackOverflow (http://stackoverflow.com/questions/36040830/casperjs-why-does-my-url-change-to-aboutblank-when-my-page-is-loaded), but it just doesn't go.

**Comment by istr** [comment]
Confirmed to occur with PhantomJS 2.1.1 only. Versions up to and including PhantomJS 2.0.1 work like expected.

**Comment by mlalpho** [comment]
I've noticed this, but, it doesn't seem to affect anything on my end. What I assume is that PhantomJS "navigates" to "about:blank" after the final navigation step set by casper has completed. Which is definitely curious, but, not a showstopper.....

What result does `this.test.assertTitle('Google', 'Welcome to Google');` give for you @rhires ?

**Comment by rhires** [comment]
@mlalpho  I guess the script never gets that far, as the log above shows. The problem is that I cannot navigate beyond this point. I get stuck. It's a showstopper for me, unfortunately. 

@istr  Should I report this issue to the phantomjs repo?

**Comment by mlalpho** [comment]
@rhires, how are you executing this script?

this.test shouldn't be accessible unless you are running `casperjs test script.js`

It would seem that casper is just silently failing as this works just fine:
```

**Comment by mlalpho** [comment]
I would say that the silent failure is a showstopper, but, is completely separate from the `about:blank` issue. Might be worth pinging PhantomJS about that. ;)

**Comment by mlalpho** [comment]
I'm wondering why this error isn't being thrown in this situation:
https://github.com/casperjs/casperjs/blob/9eb2b11954820a5b3e7854017ecb419a504b15fb/modules/casper.js#L178
@istr any clue?

**Comment by rhires** [comment]
```casperjs test.js``` is my commandline. Commenting out the this.testline does not alter the output. 

I ran your "runs just fine" script and got this as my result:

```
[debug] [phantom] Successfully injected Casper client-side utilities                                                                                                
[debug] [phantom] start page is loaded                                                                                                                              
[info] [phantom] Step anonymous 3/3 https://www.google.fr/ (HTTP 200)   
Page Loaded                                                                                                                                                         
[info] [phantom] Step anonymous 3/3: done in 451ms.                                                                                                                 
[info] [phantom] Done 3 steps in 485ms                                                                                                                              
[debug] [phantom] Navigation requested: url=about:blank, type=Other, willNavigate=true, isMainFrame=true                                                            
[debug] [phantom] url changed to "about:blank"                                                                                                                      
                                                                                                                                                                    
c:\Code\Projects>/node_modules/casperjs/bin/casperjs S3.js                                                                                                          
[info] [phantom] Starting...                                                                                                                                        
[info] [phantom] Running suite: 5 steps                                                                                                                             
[debug] [phantom] opening url: https://www.google.fr/, HTTP GET                                                                                                     
[debug] [phantom] Navigation requested: url=https://www.google.fr/, type=Other, willNavigate=true, isMainFrame=true                                                 
[debug] [phantom] url changed to "https://www.google.fr/"                                                                                                           
[debug] [phantom] Successfully injected Casper client-side utilities                                                                                                
[debug] [phantom] start page is loaded                                                                                                                              
[info] [phantom] Step anonymous 3/5 https://www.google.fr/ (HTTP 200)                                                                                               
Page Loaded                                                                                                                                                         
[info] [phantom] Step anonymous 3/5: done in 428ms.                                                                                                                 
[debug] [phantom] opening url: https://www.google.com/, HTTP GET                                                                                                    
[debug] [phantom] Navigation requested: url=https://www.google.com/, type=Other, willNavigate=true, isMainFrame=true                                                
[debug] [phantom] url changed to "https://www.google.com/"                          
[debug] [phantom] Successfully injected Casper client-side utilities                                                                                                
[info] [phantom] Step anonymous 5/5 https://www.google.com/ (HTTP 200)                
Another Page Loaded                                                                                                                                                 
[info] [phantom] Step anonymous 5/5: done in 698ms.                                                                                                                 
[info] [phantom] Done 5 steps in 717ms                                                                                                                              
[debug] [phantom] Navigation requested: url=about:blank, type=Other, willNavigate=true, isMainFrame=true                                                            
[debug] [phantom] url changed to "about:blank" 
```

Ok, so what is kicking off the "Navigation requested" ? How would I find out?

**Comment by mlalpho** [comment]
commenting out the line in your current script wouldn't alter the output, but, would allow you to navigate after the google.fr step.

The final "Navigation requested" for "about:blank" is a mystery, at this point, and according to @istr's testing it's only found when running phantomjs > 2.0.1

As you can see in the "runs just fine" sample, after commenting out the call to this.test, you are able to navigate to more than one page. 

I think the title of this issue is misleading as this has got nothing to do with being able to access https sites.

I'd recommend that you read the documentation on the Tester module if you wish to use it. Scripts executed by `casperjs test.js` and `casperjs test test.js` are not interchangeable and require somewhat different syntax to write. Oneach example being that you can't call Casper.create from within the "test" execution environment.

**Comment by rchatham** [comment]
Are there any solutions for the `about:blank` issue? I am finding the same issue in code that I am running.

**Comment by edmcman** [comment]
Me too.  Wondering how this is still open...

**Comment by mlalpho** [comment]
Can't seem to fix this from the casperjs side. PhantomJS outputs this:
```
[debug] [phantom] Navigation requested: url=about:blank, type=Other, willNavigate=true, isMainFrame=true
[debug] [phantom] url changed to "about:blank"
```
whenever phatom.exit() is called. (whether you call casper.exit() or phantom.exit() directly)

Is this preventing scripts from working? I haven't noticed any errors other than the extraneous debug output. @rchatham @edmcman 

The original problem the OP had was due to their incorrect running of the casper script. (`casperjs script.js` vs. `casperjs test script.js`)

```
$ casperjs --version
1.1.1
$ phantomjs --version
2.1.1
```

**Comment by edmcman** [comment]
@mlalpho I thought that the issue was preventing scripts from working.  I'll try to verify.

**Comment by kensoh** [comment]
Hi @mlalpho looks like the point you highlighted has been fixed. When I try to run this.test.assertX without running with test option, CasperJS throws error as expected - **CasperError: casper.test property is only available using the \`casperjs test\` command**

```
I'm wondering why this error isn't being thrown in this situation:
https://github.com/casperjs/casperjs/blob/9eb2b11954820a5b3e7854017ecb419a504b15fb/modules/casper.js#L178
@istr any clue?
```

**Comment by kensoh** [comment]
@rhires @edmcman @rchatham closing this issue for now as @mlalpho identified the root cause of error for this issue, the use of casperjs test script_name.js for running test scripts. In any case, do ping if you still have issues or would like to discuss more on the topic. Thanks!

**Comment by JefferyHus** [comment]
I have the same issue here but just when I set proxy options:

```javascript
casperjs caspertest_script.js --proxy=185.122.170.226:80 --proxy-auth=jon:jondoe20 --proxy-type=http
```

```console
[info] [phantom] Starting...
[info] [phantom] Running suite: 3 steps
[debug] [phantom] Successfully injected Casper client-side utilities
[debug] [phantom] opening url: https://www.google.com/flights/?hl=en_US#search;f
=BKK;t=SIN;d=2017-07-17;r=2017-07-24;sel=BKKSIN03K510;a=3k;s=0, HTTP GET
[debug] [phantom] Navigation requested: url=https://www.google.com/flights/?hl=e
n_US#search;f=BKK;t=SIN;d=2017-07-17;r=2017-07-24;sel=BKKSIN03K510;a=3k;s=0, typ
e=Other, willNavigate=true, isMainFrame=true
[info] [phantom] Step anonymous 1/3: done in 47ms.
[debug] [phantom] Navigation requested: url=https://www.google.nl/flights/?hl=en
_US, type=Other, willNavigate=true, isMainFrame=true
[debug] [phantom] url changed to "https://www.google.nl/flights/?hl=en_US"
[debug] [phantom] Successfully injected Casper client-side utilities
[info] [phantom] Step _step 2/3 https://www.google.nl/flights/?hl=en_US (HTTP 20
0)
[info] [phantom] Step _step 2/3: done in 5140ms.
[debug] [phantom] Navigation requested: url=about:blank, type=Other, willNavigat
e=true, isMainFrame=false
[debug] [phantom] Navigation requested: url=about:blank, type=Other, willNavigat
e=true, isMainFrame=false
[debug] [phantom] Navigation requested: url=about:blank, type=Other, willNavigat
e=true, isMainFrame=false
[warning] [phantom] Casper.waitFor() timeout
timeout occured
[info] [phantom] Step anonymous 3/3 https://www.google.nl/flights/?hl=en_US (HTT
P 200)
[CURRENT_URL]https://www.google.nl/flights/?hl=en_US
[info] [phantom] Step anonymous 3/3: done in 10240ms.
[info] [phantom] Done 3 steps in 10260ms
[debug] [phantom] Navigation requested: url=about:blank, type=Other, willNavigat
e=true, isMainFrame=true
[debug] [phantom] url changed to "about:blank"
```

**Comment by kensoh** [comment]
Thanks @JefferyHus for sharing your findings! This looks like expected behavior if your proxy is based in Netherlands I believe Google will automatically redirect to the appropriate country-specific URL. I notice this behavior and usually use Yahoo if I want to test and don't want the URL to redirect. For example https://ca.yahoo.com I am able to view it as that instead of being redirect to Singapore or some European URL.

Also, I think I understand the issue raised by you guys here in the thread now. You are saying why there is a final navigation to about:blank at the end of the automation execution? I've been experiencing this for a while but never thought much about it and just ignored it. But from Ingo's comment it sounds like some issue that was introduced in PhantomJS v2.1.1.

Can you share more what is the problem if such final step to go to about:blank is executed? For me, if I don't run in debug mode it will be transparent to me and what I am trying to do in the automation steps. Reopening the issue with updated title.

```
[info] [phantom] Step anonymous 3/3: done in 3651ms.
[info] [phantom] Done 3 steps in 3652ms
[debug] [phantom] Navigation requested: url=about:blank, type=Other, willNavigate=true, isMainFrame=true
[debug] [phantom] url changed to "about:blank"
```

**Comment by JefferyHus** [comment]
@kensoh for my case i have come to prevent redirection from google by adding `?gl=US` to the final URL, as for the `"about:blank"` issue it still remaining:
```console
[info] [phantom] Step anonymous 3/3: done in 5413ms.
[info] [phantom] Done 3 steps in 5432ms
[debug] [phantom] Navigation requested: url=about:blank, type=Other, willNavigat
e=true, isMainFrame=true
[debug] [phantom] url changed to "about:blank"
```
and it costs some time in executing the script, it is not affecting the output results, and still doing what it is asked to do, but the only concern is that it adds some response time to the execution script.

**Comment by kensoh** [comment]
I see, thanks @JefferyHus for sharing this tip and more about the about:blank issue. That's true, for large volume runs, especially tests with short iteration time, this probably adds up significantly.

I'm not expert in CasperJS internal code, otherwise I can trace to see if the issue happens over at CasperJS due to some change in PhantomJS, or if the issue inherently exists at PhantomJS.

**Comment by Tynael** [comment]
I am also having this issue...

**Comment by kensoh** [comment]
Bumping up and copying maintainer @jefleponot

**Comment by aclindsa** [comment]
For anyone still having the `about:blank` issue, it was resolved for me by upgrading to casperjs 1.1.4-1
